File: prettier-plugin-csharpier/src/Printer/Types/VariableDeclarator.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { HasIdentifier, printIdentifier, printSyntaxToken, SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { BracketedArgumentListNode, printBracketedArgumentList } from "./BracketedArgumentList";
import { EqualsValueClauseNode, printEqualsValueClause } from "./EqualsValueClause";

export interface VariableDeclaratorNode extends SyntaxTreeNode<"VariableDeclarator"> {
    identifier: SyntaxToken;
    argumentList?: BracketedArgumentListNode;
    initializer?: EqualsValueClauseNode;
}

export const printVariableDeclarator: PrintMethod<VariableDeclaratorNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [];
    parts.push(printIdentifier(node));
    if (node.argumentList) {
        parts.push(path.call(o => printBracketedArgumentList(o, options, print), "argumentList"));
    }
    if (node.initializer) {
        parts.push(path.call(o => printEqualsValueClause(o, options, print), "initializer"));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/VariableDeclarator.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class VariableDeclarator
{
    public static Doc Print(VariableDeclaratorSyntax node, PrintingContext context)
    {
        var docs = new ValueListBuilder<Doc>([null, null, null]);
        docs.Append(Token.Print(node.Identifier, context));

        if (node.ArgumentList != null)
        {
            docs.Append(BracketedArgumentList.Print(node.ArgumentList, context));
        }
        if (node.Initializer != null)
        {
            docs.Append(EqualsValueClause.Print(node.Initializer, context));
        }
        return Doc.Concat(ref docs);
    }
}
